File: Zeze/Raft/RocksRaft/TableKey.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Zeze.Serialize;

namespace Zeze.Raft.RocksRaft
{
    public class TableKey : IComparable<TableKey>
    {
        public string Name;
        public object Key;

        public TableKey()
        {

        }

        public TableKey(string name, object key)
        {
            Name = name;
            Key = key;
        }

        public override string ToString()
        {
            return $"({Name},{Key})";
        }

        public override int GetHashCode()
        {
            const int prime = 31;
            int result = 17;
            result = prime * result + Name.GetHashCode();
            result = prime * result + Key.GetHashCode();
            return result;
        }

        public override bool Equals(object obj)
        {
            if (this == obj)
                return true;

            if (obj is TableKey another)
            {
                return Name.Equals(another.Name) && Key.Equals(another.Key);
            }
            return false;
        }

        public int CompareTo(TableKey other)
        {
            int c = this.Name.CompareTo(other.Name);
            if (c != 0)
            {
                return c;
            }
            return Comparer<IComparable>.Default.Compare((IComparable)Key, (IComparable)other.Key);
        }
    }
}

File: ZezeJava/ZezeJava/src/main/java/Zeze/Raft/RocksRaft/TableKey.java
package Zeze.Raft.RocksRaft;

public final class TableKey implements Comparable<TableKey> {
	public String name;
	public Object key;

	public TableKey() {
	}

	public TableKey(String name, Object key) {
		this.name = name;
		this.key = key;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 17;
		result = prime * result + name.hashCode();
		result = prime * result + key.hashCode();
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;

		if (obj instanceof TableKey) {
			var another = (TableKey)obj;
			return name.equals(another.name) && key.equals(another.key);
		}
		return false;
	}

	@SuppressWarnings("unchecked")
	@Override
	public int compareTo(TableKey other) {
		int c = name.compareTo(other.name);
		if (c != 0)
			return c;
		return ((Comparable<Object>)key).compareTo(other.key);
	}

	@Override
	public String toString() {
		return String.format("(%s,%s)", name, key);
	}
}
